frame/elections-phragmen/src/lib.rs

```diff
@@ -289,210 +324,160 @@ decl_error! {
+		///
+		/// ### Warning
+		///
+		/// It is the responsibility of the caller to **NOT** place all of their balance into the
+		/// lock and keep some for further transactions.
 		///
 		/// # <weight>
-		/// Base weight: 47.93 µs
-		/// State reads:
-		/// 	- Candidates.len() + Members.len() + RunnersUp.len()
-		/// 	- Voting (is_voter)
-		/// 	- Lock
-		/// 	- [AccountBalance(who) (unreserve + total_balance)]
-		/// State writes:
-		/// 	- Voting
-		/// 	- Lock
-		/// 	- [AccountBalance(who) (unreserve -- only when creating a new voter)]
+		/// we consider the common case of placing more votes. In other two case, we refund.
 		/// # </weight>
-		#[weight = T::WeightInfo::vote(votes.len() as u32)]
+		#[weight =
+			T::WeightInfo::vote_more(votes.len() as u32)
+			.max(T::WeightInfo::vote_less(votes.len() as u32))
+			.max(T::WeightInfo::vote_equal(votes.len() as u32))
+		]
 		fn vote(
 			origin,
 			votes: Vec<T::AccountId>,
 			#[compact] value: BalanceOf<T>,
 		) {
 			let who = ensure_signed(origin)?;
 
+			// votes should not be empty and more than `MAXIMUM_VOTE` in any case.
 			ensure!(votes.len() <= MAXIMUM_VOTE, Error::<T>::MaximumVotesExceeded);
 			ensure!(!votes.is_empty(), Error::<T>::NoVotes);
 
 			let candidates_count = <Candidates<T>>::decode_len().unwrap_or(0);
 			let members_count = <Members<T>>::decode_len().unwrap_or(0);
 			let runners_up_count = <RunnersUp<T>>::decode_len().unwrap_or(0);
 
+			// can never submit a vote of there are no members, and cannot submit more votes than
+			// all potential vote targets.
 			// addition is valid: candidates, members and runners-up will never overlap.
 			let allowed_votes = candidates_count + members_count + runners_up_count;
-
 			ensure!(!allowed_votes.is_zero(), Error::<T>::UnableToVote);
 			ensure!(votes.len() <= allowed_votes, Error::<T>::TooManyVotes);
 
 			ensure!(value > T::Currency::minimum_balance(), Error::<T>::LowBalance);
 
-			// first time voter. Reserve bond.
-			if !Self::is_voter(&who) {
-				T::Currency::reserve(&who, T::VotingBond::get())
-					.map_err(|_| Error::<T>::UnableToPayBond)?;
-			}
+			// Reserve bond.
+			let new_deposit = Self::deposit_of(votes.len());
+			let Voter { deposit: old_deposit, .. } = <Voting<T>>::get(&who);
+			match new_deposit.cmp(&old_deposit) {
+				Ordering::Greater => {
+					// Must reserve a bit more.
+					let to_reserve = new_deposit - old_deposit;
+					T::Currency::reserve(&who, to_reserve)
+						.map_err(|_| Error::<T>::UnableToPayBond)?;
+				},
+				Ordering::Equal => {},
+				Ordering::Less => {
+					// Must unreserve a bit.
+					let to_unreserve = old_deposit - new_deposit;
+					let _remainder = T::Currency::unreserve(&who, to_unreserve);
+					debug_assert!(_remainder.is_zero());
+				},
+			};
 
 			// Amount to be locked up.
-			let locked_balance = value.min(T::Currency::total_balance(&who));
+			let locked_stake = value.min(T::Currency::total_balance(&who));
```

Comment:
I want someone to correct me here: Is this sane? Essentially, you can still lock up tokens that you have reserved in the system for voting, and you can seemingly use them for transaction fees. I would change that to `WithdrawReasons::all()` cc @shawntabrizi @thiolliere

Reply:
I can't really see why locking only for transaction-payment reason here. But I can't really say about those economic constraints/incentives.

Reply:
> I can't really see why locking only for transaction-payment reason here.

FYI we do the opposite, locking except for transaction payment, so you can pay for your fees with the vote stake that you've locked. Maybe @gavofyork can chime in? Else I would make this `::all()`, and change the doc to demonstrate.

Reply:
As promised, I updated this.

Reply:
Transaction fees mostly go into the system treasury (the same place as any slashes) so I'd generally consider them safe to spend out of a democracy vote (unlike e.g. tips). But it's mostly just to avoid the papercut of locking all your funds in a vote and then not having the funds to pay the fee to unlock them. It's probably fine as wallets/infra begins to understand stuff like this and automatically keeps a little back from being locked, to pay for any future transaction fees.